src/plugins/lastfm.ts

```diff
@@ -0,0 +1,185 @@
+}
+
+function setActivity(activity?: Activity) {
+    rpcServer.dispatch({
+        type: "LOCAL_ACTIVITY_UPDATE",
+        activity: activity && Object.assign(
+            activity, { flags: ActivityFlag.INSTANCE, type: ActivityType.PLAYING }
+        )
+    });
+}
+
+export default definePlugin({
+    name: "LastFMRichPresence",
+    description: "Little plugin for Last.fm rich presence",
+    authors: [
+        {
+            name: "dzshn",
+            id: 310449948011528192n,
+        },
+    ],
+    options: {
+        username: {
+            description: "last.fm username",
+            type: OptionType.STRING,
+            placeholder: "dimdendev",
+        },
+        apiKey: {
+            description: "last.fm api key",
+            type: OptionType.STRING,
+            placeholder: "686f6765686f6765686f6765686f6765",
+        },
+        hideWithSpotify: {
+            description: "hide last.fm presence if spotify is running",
+            type: OptionType.BOOLEAN,
+            default: true,
+        }
+    },
+    start() {
+        this.settings = Settings.plugins.LastFMRichPresence;
+
+        setInterval(() => { this.updatePresence(); }, 16000);
+    },
+    stop() {},
+    async fetchTrackData(): Promise<TrackData | null> {
+        if (!this.settings.username || !this.settings.apiKey) return null;
+
+        const response = await fetch(`https://ws.audioscrobbler.com/2.0/?method=user.getrecenttracks&api_key=${this.settings.apiKey}&user=${this.settings.username}&limit=1&format=json`);
+        const trackData = (await response.json()).recenttracks.track[0];
+
+        if (!trackData["@attr"]?.nowplaying) return null;
+
+        // why does the json api have xml structure
+        return {
+            name: trackData.name || "Unknown",
+            album: trackData.album["#text"],
+            artist: trackData.artist["#text"] || "Unknown",
+            url: trackData.url,
+            imageUrl: (trackData.image || []).filter(x => x.size === "large")[0]?.["#text"]
+        };
+    },
+    async updatePresence() {
+        if (this.settings.hideWithSpotify) {
+            for (const activity of presenceStore.getActivities()) {
+                if (activity.type === ActivityType.LISTENING) {
+                    // there is already music status (probably only spotify can do this currently)
+                    setActivity();
+                    return;
+                }
+            }
+        }
+
+        const trackData = await this.fetchTrackData();
+
+        if (!trackData) {
+            setActivity();
+            return;
+        }
+
+        const hideAlbumName = !trackData.album || trackData.album === trackData.name;
+
+        let assets: ActivityAssets;
+        if (trackData.imageUrl) {
+            assets = {
+                large_image: await getApplicationAsset(trackData.imageUrl),
+                large_text: trackData.name,
+                small_image: await getApplicationAsset("lastfm-small"),
```

Comment:
probably would be best to cache these
```suggestion
                small_image: await getSmallImage(),
```

then on top level add
```js
const getSmallImage = makeLazy(() => getApplicationAsset("lastfm-small"))
```

Reply:
leaving it as-is since it's already cached by that function